Changed the guzzle client for the symfony httpclient

bramtervoort opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | yes
| Fixed tickets | -
| Related issues/PRs | -
| License | MIT
| Documentation PR | -

#### What's in this PR?

Changed the guzzle client for the httpclient of symfony in the Locationbundle

#### Why?

It seems easier to update a standard symfony component and now the guzzle client could be dropped as a dependency for the LocationBundle. 

#### BC Breaks/Deprecations?

The constructor of the `NominatimGeolocator` and `GoogleGeolocator` requires a `Symfony\Contracts\HttpClient\HttpClientInterface` for the `$client` argument

Constructing `NominatimGeolocator` and `GoogleGeolocator` with the previous `GuzzleHttp\ClientInterface` is deprecated. 

updated UPGRADE.md

Commit:
Changed the guzzle client for the httpclient of symfony

It seems easier to update a standard symfony component
and now the guzzle client could be dropped as a dependency.

bramtervoort commented:
My first ever pull request. I hope i did this right :D if not then ill be happy to change it or discard my changes. Thank you

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Resources/config/geolocator.xml:
```diff
@@ -2,26 +2,22 @@
         <!-- * Nominatim * -->
         <service id="sulu_location.geolocator.service.nominatim" class="%sulu_location.geolocator.nominatim.class%">
-            <argument type="service" id="sulu_location.geolocator.guzzle.client" />
+            <argument type="service" id="Symfony\Contracts\HttpClient\HttpClientInterface" />
```
Bundles should not use aliases, and so in this case use `http_client` here as service definition.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Resources/config/geolocator.xml:
```diff
@@ -2,26 +2,22 @@
         <!-- * Google * -->
         <service id="sulu_location.geolocator.service.google" class="%sulu_location.geolocator.google.class%">
-            <argument type="service" id="sulu_location.geolocator.guzzle.client" />
+            <argument type="service" id="Symfony\Contracts\HttpClient\HttpClientInterface" />
```
Bundles should not use aliases, and so in this case use `http_client` here as service definition.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Geolocator/Service/NominatimGeolocator.php:
```diff
@@ -39,16 +40,19 @@ class NominatimGeolocator implements GeolocatorInterface
-    {
-        $this->client = $client;
+    public function __construct(
+        HttpClientInterface $httpClient,
```
This is a BC break. We need here handle that somebody maybe inject the guzzle client.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Geolocator/Service/NominatimGeolocator.php:
```diff
@@ -25,9 +26,9 @@
 {
     /**
-     * @var ClientInterface
+     * @var HttpClientInterface
```
This still could be both `ClientInterface|HttpClientInterface` as our bc promise.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Geolocator/Service/NominatimGeolocator.php:
```diff
@@ -39,16 +40,19 @@ class NominatimGeolocator implements GeolocatorInterface
+        string $baseUrl,
+        string $key
+    ) {
+        $this->httpClient = $httpClient;
```
It should be throw a deprecation message if its guzzle client here.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Geolocator/Service/NominatimGeolocator.php:
```diff
@@ -25,9 +26,9 @@
+     * @var HttpClientInterface
      */
-    protected $client;
+    protected $httpClient;
```
would keep the variable as `client`

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Geolocator/Service/NominatimGeolocator.php:
```diff
@@ -72,7 +76,7 @@ public function locate(string $query): GeolocatorResponse
         }
 
-        $results = \json_decode($response->getBody(), true);
+        $results = $response->toArray();
```
we need here handle if its `Guzzle` or `Symfony` client because of our bc promise.

alexander-schranz reviewed (changes requested):
@bramtervoort thx for the PR. I think it is a good idea to use here also the symfony/http-client. Because of our bc promise we need still handle both if the Locator service is constructed with guzzle or symfony client. So we can remove in the next major release then the guzzle implementation totally.

bramtervoort commented:
Thank you for your review, indeed I broke bc without realizing. I will refactor a little to change it.

Commit:
Fixes bc for guzzle to symfony http client switch of the LocationBundle

Readded support for the guzzle client and added the test, also added the
bc breaking changes to the documentation.

bramtervoort commented:
I fixed the bc breaking changes to a nice deprecation message, updated the UPGRADE.md and added http_client instead of the alias.

I understand that its the holidays so no rush in merging this. I just enjoy making some contributions in my weeks off. I'd be happy to make more changes or tests if needed.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Resources/config/geolocator.xml:
```diff
@@ -2,26 +2,22 @@
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
     <parameters>
-        <parameter key="sulu_location.guzzle.client.class">GuzzleHttp\Client</parameter>
```
from bc we can not remove this parameter

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Resources/config/geolocator.xml:
```diff
@@ -2,26 +2,22 @@
 
     <services>
-        <!-- Guzzle client -->
-        <service id="sulu_location.geolocator.guzzle.client" class="%sulu_location.guzzle.client.class%"/>
```
also this service should still exist as maybe somebody did use it.

bramtervoort replied:
oo indeed and it should be deprecated to.

alexander-schranz reviewed (changes requested):
Just some little change request for the service definition to keep backward compatibility.

Commit:
Fixes bc `sulu_location.geolocator.guzzle.client` and parameter `sulu_location.guzzle.client.class`

bramtervoort commented:
fixed the issues i think. happy holiday's if stuff needs more changes then i'd be hapy to do that ofcourse but holliday's first.

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Resources/config/geolocator.xml:
```diff
@@ -9,19 +9,22 @@
+        <service id="sulu_location.geolocator.guzzle.client" alias="sulu_location.geolocator.guzzle.client_deprecated">
+            <deprecated />
+        </service>
+        <service id="sulu_location.geolocator.guzzle.client_deprecated" class="%sulu_location.guzzle.client.class%"/>
```
Do we need this service definition can we not just set the class on the above service definition which has the `<deprecated />` or can only aliases be deprecated?

bramtervoort replied:
Happy new year. O yes your right i can also depricate services. I read the doc's wrong i think 
https://symfony.com/blog/new-in-symfony-4-3-deprecating-service-aliases

but this is been here for services all along https://symfony.com/blog/new-in-symfony-2-8-deprecated-service-definitions 

So nice to learn all this. Thank you for your time in helping me with this.

bramtervoort replied:
lol wou i should have checked that better. :D

alexander-schranz replied:
Thx also a Happy new year!. Good to hear that you also learn from this PR :).

alexander-schranz reviewed (changes requested):
@bramtervoort Thx, I'm also now back from holidays. Hope yours where also fine. Just some comment I'm not sure if we really need 2 service definitions.

bramtervoort committed:
Fixed depricating guzzle client instead of alias

alexander-schranz commented on src/Sulu/Bundle/LocationBundle/Resources/config/geolocator.xml:
```diff
@@ -9,19 +9,21 @@
     <services>
         <!-- Guzzle client -->
-        <service id="sulu_location.geolocator.guzzle.client" class="%sulu_location.guzzle.client.class%"/>
+        <service id=""sulu_location.geolocator.guzzle.client" class="%sulu_location.guzzle.client.class%">
```
```suggestion
        <service id="sulu_location.geolocator.guzzle.client" class="%sulu_location.guzzle.client.class%">
```

There is a acccidently `""`

bramtervoort replied:
yes thout to edit that very quickly. Always a bad idear to push unchecked code. fixed this now.

Commit:
Fixed xml typo for guzzle client

bramtervoort commented:
@alexander-schranz Thank you for approving. Do you need me to rebase and compress my commits into one commit? I still see a cs-fixer failing but that does not seem to be a file I changed.

alexander-schranz commented:
@bramtervoort CS Fixer was fixed by @nnatter in a seperate PR as we always need todo this on the lowest supported branch first (1.6). So you can ignore that. Commits are squashed by us  using the Github Squash & Merge feature so no rebase is needed. I'm just waiting for appveyor to be sure that it runs well on windows :). When that is finished I will merge it.

alexander-schranz commented:
@bramtervoort Thank you for your effort!

niklasnatter commented:
@bramtervoort Nice work - thank you! 🙂
